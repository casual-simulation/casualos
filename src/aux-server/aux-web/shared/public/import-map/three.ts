export * from '@casual-simulation/three';

// Export the URL for the import map
export const url = import.meta.url