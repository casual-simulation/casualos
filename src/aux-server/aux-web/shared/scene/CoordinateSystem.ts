import {
    Vector3,
    MathUtils as ThreeMath,
    Matrix4,
} from '@casual-simulation/three';

/**
 * Defines the possible coordinate systems.
 */
export enum CoordinateSystem {
    /**
     * Specifies that the coordinate system uses Z for the up axis.
     * This is what AUX uses.
     */
    Z_UP,

    /**
     * Specifies that the coordinate system uses Y for the up axis.
     * This is what the Three.js scenes use by default.
     */
    Y_UP,
}

export interface CoordinateTransformer {
    (pos: { x: number; y: number; z: number }): Matrix4;
}

/**
 * Converts the given latitude and longitude point to a cartesian point.
 * @param radius The radius of the earth.
 * @param point The 3D point representing latitude (X), longitude (Y), and altitude (Z).
 */
export function latLonToCartesian(radius: number, point: Vector3): Vector3 {
    const lat = ThreeMath.degToRad(point.x);
    const lon = ThreeMath.degToRad(point.y);
    const altitude = point.z;
    const heightfactor = radius + altitude;
    const x = heightfactor * Math.cos(lat) * Math.cos(lon);
    const y = heightfactor * Math.cos(lat) * Math.sin(lon);
    const z = heightfactor * Math.sin(lat);
    return new Vector3(x, y, z);
}

/**
 * Converts the given cartesian point to a latitude and longitude point.
 * @param radius The radius of the earth.
 * @param point The 3D point representing the cartesian coordinate.
 */
export function cartesianToLatLon(radius: number, point: Vector3): Vector3 {
    const normalized = point.clone().normalize().multiplyScalar(radius);
    const lon = Math.atan2(normalized.y, normalized.x);
    const lat = Math.asin(normalized.z / radius);
    const distanceFromRadius = point.distanceTo(normalized);
    return new Vector3(
        ThreeMath.radToDeg(lat),
        ThreeMath.radToDeg(lon),
        distanceFromRadius
    );
}
