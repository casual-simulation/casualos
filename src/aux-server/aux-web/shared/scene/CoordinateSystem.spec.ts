import { Vector3 } from '@casual-simulation/three';
import { cartesianToLatLon, latLonToCartesian } from './CoordinateSystem';

describe('CoordinateSystem', () => {
    describe('latLonToCartesian()', () => {
        it('should return a Vector3 for the given latitude and longitude and altitude', () => {
            let input = new Vector3(9, 12, 15);
            let output = latLonToCartesian(1, input);
            expect(output).toMatchSnapshot();
        });
    });

    describe('cartesianToLatLon()', () => {
        it('should return a Vector3 for the given latitude and longitude and altitude', () => {
            let input = new Vector3(
                15.457679690014206,
                3.285631246305953,
                2.502951440643694
            );
            let output = cartesianToLatLon(1, input);

            expect(output.x).toBeCloseTo(9);
            expect(output.y).toBeCloseTo(12);
            expect(output.z).toBeCloseTo(15);
        });
    });
});
