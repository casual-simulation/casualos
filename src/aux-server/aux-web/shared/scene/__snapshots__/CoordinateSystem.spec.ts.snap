// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoordinateSystem latLonToCartesian() should return a Vector3 for the given latitude and longitude and altitude 1`] = `
Vector3 {
  "x": 15.457679690014206,
  "y": 3.285631246305953,
  "z": 2.502951440643694,
}
`;
